src/lib/ImmediateList.re

```diff
@@ -0,0 +1,29 @@
+type glRenderFunction = (int, float) => unit;
+
+let additionalRowsToRender = 1;
+
+let render =
+    (~scrollY=0., ~rowHeight, ~height, ~count, ~render: glRenderFunction, ()) => {
+  let rowsToRender = rowHeight > 0. ? int_of_float(height /. rowHeight) : 0;
+  let startRowOffset =
+    rowHeight > 0. ? int_of_float(scrollY /. rowHeight) : 0;
+  let pixelOffset = mod_float(scrollY, rowHeight);
+
+  let i = ref(max(startRowOffset - additionalRowsToRender, 0));
+
+  let len = count;
+
+  while (i^ < rowsToRender
+         + additionalRowsToRender
+         + startRowOffset
+         && i^ < len) {
+    let item = i^;
+    let rowOffset = float_of_int(item - startRowOffset) *. rowHeight;
+
+    let top = rowOffset -. pixelOffset;
+
+    render(item, top);
+
+    incr(i);
+  };
+};
```

Comment:
You might want to copy the implementation from FlatList instead. It's much less imperative and a lot nicer in my extremely humble opinion.

Reply:
I pulled this from the `ImmediateList` in Onivim (https://github.com/onivim/oni2/blob/master/src/Feature/Editor/ImmediateList.re)

There's a difference in semantics - `<FlatList />` operates at the _component_ level whereas `ImmediateList` is for the immediate-mode rendering (ie, inside a canvas). I'd prefer to render the lines in the context of the canvas. 

I don't see a benefit to switching the rendering pipeline to go back to components, just to enable switching to `<FlatList />`. Just seems like extra work, with a potential performance cost (more expensive reconciliation)

Reply:
The `FlatList.render` implementation isn't inherently component+based. It just happens to wrap it in a component in this case instead of passing `offset` to the render function. But if you remove that I'm pretty sure it'll work the same.

Reply:
I see, thanks for clarifying, I missed the `render` part.

I'm open to the change, however what we have with the `ImmediateList` works (and it is tested in Onivim). I'm not convinced this spending the additional time to convert it, along with the risk of potentially breaking/regressing what I tested so far, is justified. 

Unfortunately without the luxury of infinite time - we have to be pragmatic and find  balance between perfection and shipping - and I don't think this strikes the right balance at this time.